server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -476,11 +476,26 @@ public long ramBytesUsed() {
     }
 
     /**
-     * Returns how much RAM would be freed up by refreshing. This is the RAM usage of the current version map. It doesn't include tombstones
-     * since they don't get cleared on refresh, nor the old version map that is being reclaimed.
+     * Returns how much RAM is used by refresh. This is the RAM usage of the current and old version maps, and the RAM usage of the
+     * archive, if any.
      */
     long ramBytesUsedForRefresh() {
-        return maps.current.ramBytesUsed.get();
+        return maps.ramBytesUsed() + archive.getMemoryBytesUsed();
```

Comment:
This change affects also Stateful. If necessary, i could separate it out, or only do that change for Stateless. To me it seems that when [calculating the current shard heap usages](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java#L366-L372), we should consider both map usages and then subtract the memory usage that is being freed.

Reply:
I've separated the stateful change to a new PR https://github.com/elastic/elasticsearch/pull/104122. I think that will make it easier to review this one.